File: modules/libs/core/flashcards/review/grid.go
package review

import "math"

// LeastCeiling is the shortest day a curve of minutes runs to.
const LeastCeiling = 60

// spread is up to as many places as are wanted, evenly over the days, and
// always both ends of them.
func spread(days, places int) []int {
	if days <= places {
		out := make([]int, days)
		for i := range out {
			out[i] = i
		}
		return out
	}
	out := make([]int, places)
	for i := range out {
		out[i] = int(math.Round(float64(i) * float64(days-1) / float64(places-1)))
	}
	return out
}

// setNearestStep puts one place of the range on the grid, in place of the place
// nearest it. The two ends stand: a range begins tomorrow and reaches as far as
// it reaches, whatever day the file names.
//
// The point under the place is worked out for the day the preset aims at.
func setNearestStep(steps []int, at int) []int {
	if at < 0 || len(steps) < 3 {
		return steps
	}
	near := 1
	for i := 2; i < len(steps)-1; i++ {
		if abs(steps[i]-at) < abs(steps[near]-at) {
			near = i
		}
	}
	if at > steps[0] && at < steps[len(steps)-1] {
		steps[near] = at
	}
	return steps
}

// snap puts the value the preset holds on the grid, in place of the place of
// it nearest that value, so what is drawn under the place is drawn for the
// setting the person is standing at.
//
// First and last are the places a value may take: a range whose ends say what
// the setting may be at all keeps them.
func snap(grid []float64, value float64, first, last int) {
	if first < 0 || last >= len(grid) || first > last {
		return
	}
	at := first
	for i := first; i <= last; i++ {
		if math.Abs(grid[i]-value) < math.Abs(grid[at]-value) {
			at = i
		}
	}
	grid[at] = value
}

// abs is how far a whole number stands from nothing.
func abs(one int) int {
	if one < 0 {
		return -one
	}
	return one
}

// ceiling is how far a curve of minutes runs: twice what carrying the whole
// load costs, and never less than a short day or more than a day holds.
func ceiling(load, keeping float64) float64 {
	top := math.Ceil(2 * math.Max(load, keeping))
	return math.Min(math.Max(top, LeastCeiling), MinutesADayBounds.Most)
}

// nearest is the place of the grid a value falls at, and -1 for a value outside
// it.
func nearest(grid []float64, value float64) int {
	if len(grid) == 0 || value < grid[0] || value > grid[len(grid)-1] {
		return -1
	}
	at := 0
	for i, one := range grid {
		if math.Abs(one-value) < math.Abs(grid[at]-value) {
			at = i
		}
	}
	return at
}

File: modules/libs/ui/src/features/heatmap/lib/grid.ts
/**
 * The room a grid of days is laid out in: how many columns fit it, and the
 * stretch of time a grid that wide draws.
 */
import { getDayName } from '@/shared/lib/day'

/** How many days stand in one column. A column is a week. */
export const ROWS = 7

/** How many weeks of what is still to come the grid keeps room for. */
export const AHEAD = 4

/** What a grid is laid out to. */
export interface HeatmapMetrics {
  /** How wide the grid may be, in pixels. */
  width: number
  /** How large one cell is drawn. */
  cell: number
  /** How much room is left between two cells. */
  gap: number
}

/**
 * How many columns fit the room there is, and how large a cell is drawn in it.
 *
 * The cell has a size of its own and the grid takes as many columns as fit, so
 * a wide window shows more weeks rather than the same weeks drawn larger, and a
 * narrow one shows fewer rather than the grid standing in the middle of empty
 * room. What is left over is spread between the cells, which keeps the grid
 * flush to both edges.
 */
export function measureGrid(metrics: HeatmapMetrics): {
  columns: number
  cell: number
  gap: number
} {
  const cell = Math.max(1, metrics.cell)
  const gap = Math.max(0, metrics.gap)
  const step = cell + gap
  if (metrics.width <= 0) return { columns: 1, cell, gap }

  const columns = Math.max(1, Math.floor((metrics.width + gap) / step))
  if (columns < 2) return { columns, cell, gap }

  // The room the cells do not take is the room between them.
  const between = Math.max(gap, (metrics.width - columns * cell) / (columns - 1))
  return { columns, cell, gap: between }
}

/**
 * The days a grid of this many columns draws, oldest first.
 *
 * The weeks behind a person run up to the one they are in, and a few weeks of
 * what is still to come stand after it, so the grid says what is coming as well
 * as what was done. Every column is a whole week.
 *
 * A day still to come holds what falls on it; a day behind holds what was
 * answered on it. Today holds what was answered, because that is the number a
 * person is adding to.
 */
/**
 * The stretch of time a grid this wide draws, oldest day first, as the names a
 * day is written under.
 *
 * A caller asks the application for the days before it knows the room it will
 * have, so it asks about the widest grid the room could hold. Every day any
 * narrower grid draws stands inside the answer.
 */
export function getStretch(metrics: HeatmapMetrics, now: Date): { from: string; to: string } {
  const { columns } = measureGrid(metrics)
  const weeks = Math.min(AHEAD, Math.max(0, columns - 1))

  // The grid running back from now: the Sunday ending the last week kept for
  // what is still to come, and the columns before it.
  const last = new Date(now)
  last.setHours(12, 0, 0, 0)
  last.setDate(last.getDate() + ((7 - weekday(last)) % 7) + weeks * ROWS)
  const behind = new Date(last)
  behind.setDate(behind.getDate() - (columns * ROWS - 1))

  // The grid opening on the week a person began in, which for a vault with
  // nothing behind it is this week, and then all of it stands ahead.
  const opens = monday(now)
  const ahead = new Date(opens)
  ahead.setDate(ahead.getDate() + columns * ROWS - 1)

  return {
    from: getDayName(behind < opens ? behind : opens),
    to: getDayName(last > ahead ? last : ahead),
  }
}

/** The Monday of the week a day stands in, which is the column it opens. */
function monday(at: Date): Date {
  const out = new Date(at)
  out.setHours(12, 0, 0, 0)
  out.setDate(out.getDate() - (weekday(out) - 1))
  return out
}

/** Monday is the first day of a week, and Sunday the seventh. */
function weekday(at: Date): number {
  const day = at.getDay()
  return day === 0 ? 7 : day
}
